cpp/src/arrow/compute/exec/aggregate_node.cc

```diff
@@ -169,17 +206,63 @@ void AggregatesToString(std::stringstream* ss, const Schema& input_schema,
   *ss << ']';
 }
 
+template <typename BatchHandler>
+Status HandleSegments(std::unique_ptr<GroupingSegmenter>& segmenter,
+                      const ExecBatch& batch, const std::vector<int>& ids,
+                      const BatchHandler& handle_batch) {
+  int64_t offset = 0;
+  ARROW_ASSIGN_OR_RAISE(auto segment_batch, batch.SelectValues(ids));
+  while (true) {
+    ARROW_ASSIGN_OR_RAISE(auto segment, segmenter->GetNextSegment(segment_batch, offset));
+    if (segment.offset >= segment_batch.length) break;
+    ARROW_RETURN_NOT_OK(handle_batch(batch, segment));
+    offset = segment.offset + segment.length;
+  }
+  return Status::OK();
+}
+
+Status GetScalarFields(std::vector<Datum>& values, const ExecBatch& input_batch,
```

Comment:
```suggestion
Status GetScalarFields(std::vector<Datum>* values, const ExecBatch& input_batch,
```
Prefer pointer over mutable reference since `GetScalarFields(&values)` is more clearly modifying `values` than `GetScalarFields(values)`.

Reply:
Do we expect this to modifying "values"? The function is called "GetScalarFields" which seems like a non-modifying method

Reply:
Changed to pointer.

> Do we expect this to modify "values"? The function is called "GetScalarFields" which seems like a non-modifying method

Yes, it modifies `values`. What function name do you think would be better?

Reply:
Renamed to `SelectConstantFields`.